### PR TITLE
[ADD] bonuses technicians

### PR DESCRIPTION
**Pull request by sbuhl**
The bonuses for technicians are calculated as follows:
For each sale, a quotation is validated and an order is generated. On this order, lines related to labor (main d'oeuvre) are added.
These labor lines have the following characteristics:
- They are products of the service type.
- The service_tracking field of these products is set to task_global_project | task_in project | project_only.
- The quantity of those product should be limited to 1 unit.
- The amount (HTVA), is equal to the labour_amount (see below)
Note: An order can contain multiple labor lines.

When the order is validated and the task(s) is/are created, one or more technicians will be assigned to the task(s).  Technicians must record the number of hours spent on the task on timesheets.
We need to retrieve the total number of hours worked (total_hours) as well as the number of hours per technician (tech1_hours, tech2_hours, etc.) in order to calculate the percentage of work per technician (tech1_percentage, tech2_percentage, etc.).

~The commission percentage granted to the technician will depend on what is recorded on the contract of the employee.
(If needed, later we could add another % on the product).~
Cost Recovery Rate Technicians (cost_recovery_rate_technicians). 
Note: previously, I wanted to add this on the contract of the technicians, but after reconsidering, I think it's better to add it on the product. As we want to keep what we pay under control.

Based on this percentage, the calculation will be as follows: The total cost recovery to give to technicians is equal to the labor amount from the SO times the cost recovery rate set on the product.
total_cost_recovery_technicans = cost_recovery_rate_technicians * labor_amount
Then this amount has to be distributed among technicians, based on their timesheet.
if technician A worked 5 hours, technician B worked 3 hours and technicians C worked 2 hours. The total cost should be slipped among them proportionnaly
technicianA_ cost_recovery_amount = total_cost_recovery_technicans / total_hours_worked * technicanA_total_hours worked
technicianB_ cost_recovery_amount = total_cost_recovery_technicans / total_hours_worked * technicanB_total_hours worked
technicianC_ cost_recovery_amount = total_cost_recovery_technicans / total_hours_worked * technicanC_total_hours worked
So each labour line has one task, and each task has at least one commission.

However, the bonus will only be "given" to the technician when the following conditions are met:

- The SO must be fully invoiced.
- All invoices must be fully paid.
- All products on the SO must be delivered.
- All labor services must be in the "Done" stage (thus, on the SO, the delivered quantity matches the ordered and invoiced quantity).

If these conditions are met (when the invoice is paid, or when the final products are delivered, or when the tasks are completed, etc.), a line must be added to the technician's vendor bill for receiving the bonus.
If there is no vendor bill yet, it needs to be created and addressed to the technician. On the line, a specific product should be used (which can be defined in the settings).
If there is an existing commission, still in draft, check if it's for the same month, else create a new bill.
One can't have commission from hibou and from labour, but if it happens it's fine, whatever if hibou vendor bill & labour vendor bill are different vendor bill.
A stat button on the invoice and on the SO should display the number of commissions related to this SO.
A stat button on the task should display the number of commissions related to this work.

In the settings, add a "Bonus" section.
Add a subsection for "Technician Bonuses," including the option to choose the product and the journal where the vendor bill will be located.

In the views, we should be able to see the technician bonuses from both the Field Services and Sales applications.

On the Invoice, there should be a "Calculate Bonus" button that will trigger the computation of the bonus.

### CODE REVIEW

File: models/bonus.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        # - be fully delivered
+        # - have its related labor service tasks done (related SOL will be
+        #   marked as delivered, so it should be covered by previous point)
+        if any([move_state != 'paid' for move_state in order.invoice_ids.mapped('payment_state')]):
+            # Not all invoices are fully paid
+            return
+        if any(line for line in order.order_line if line.product_uom_qty != line.qty_invoiced):
+            # Not all products have been invoiced
+            return
+        if any(line for line in order.order_line if line.product_uom_qty != line.qty_delivered):
```
Comment (sbuhl) [inline models/bonus.py]: @shortcut1337  Est-ce que tu peux regarder ici, uniquement pour les produits stockables ou consommables.

File: views/sale_order_views.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+                <button type="object" name="action_view_bonuses" class="oe_stat_button" icon="fa-gift"
+                        attrs="{'invisible': [('bonuses_count', '=', 0)]}" groups="account.group_account_manager">
+                    <field name="bonuses_count" widget="statinfo" string="Bonuses"/>
+                </button>
+            </xpath>
+        </field>
+    </record>
+
+    <record id="action_invoice_order_regenerate_gse_bonuses" model="ir.actions.server">
+        <field name="name">Regenerate labor bonuses</field>
```
Comment (sbuhl) [inline views/sale_order_views.xml]: "Regenerate Transport Expenses"

File: views/sale_order_views.xml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        <field name="model_id" ref="model_sale_order"/>
+        <field name="groups_id" eval="[(4, ref('account.group_account_manager'))]"/>
+        <field name="binding_model_id" ref="model_sale_order" />
+        <field name="state">code</field>
+        <field name="binding_view_types">form</field>
```
Comment (sbuhl) [inline views/sale_order_views.xml]: Could you also include it in the tree view?

File: views/sale_order_views.xml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    <record id="view_order_form_inherit_gse_bonuses" model="ir.ui.view">
+        <field name="name">sale.order.form.gse.bonuses</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form"/>
+        <field name="priority">10</field>
+        <field name="arch" type="xml">
+            <xpath expr="//button[@name='action_view_invoice']" position="before">
+                <button type="object" name="action_view_bonuses" class="oe_stat_button" icon="fa-gift"
+                        attrs="{'invisible': [('bonuses_count', '=', 0)]}" groups="account.group_account_manager">
+                    <field name="bonuses_count" widget="statinfo" string="Bonuses"/>
```
Comment (sbuhl) [inline views/sale_order_views.xml]: string="Transport Expenses"

File: views/res_config_settings_views.xml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+<odoo>
+
+    <record id="res_config_settings_view_form" model="ir.ui.view">
+        <field name="name">res.config.settings.view.form.inherit.account</field>
+        <field name="model">res.config.settings</field>
+        <field name="priority" eval="50"/>
+        <field name="inherit_id" ref="base.res_config_settings_view_form"/>
+        <field name="arch" type="xml">
+            <div name="performance" position="after">
+                <h2>Bonuses</h2>
```
Comment (sbuhl) [inline views/res_config_settings_views.xml]: Reword: Expenditures

File: views/res_config_settings_views.xml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                <div class="row mt16 o_settings_container" groups="account.group_account_manager">
+                    <div class="col-xs-12 col-md-6 o_setting_box">
+                        <div class="o_setting_left_pane"/>
+                        <div class="o_setting_right_pane">
+                            <span class ="o_form_label">Technician Bonuses</span>
```
Comment (sbuhl) [inline views/res_config_settings_views.xml]: Reword: "Technician Transport Expenses"

File: views/res_config_settings_views.xml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                            <span class="fa fa-lg fa-building-o" title="Values set here are company-specific." groups="base.group_multi_company"/>
+                            <div class="content-group">
+                                <div class="row mt16">
+                                    <label string="Journal" for="bonus_journal_id" class="col-md-3 o_light_label"/>
+                                    <field name="bonus_journal_id" domain="[('company_id', '=', company_id)]"/>
+                                </div>
+                                <div class="row">
+                                    <label string="Bonus Product" for="bonus_product_id" class="col-md-3 o_light_label"/>
```
Comment (sbuhl) [inline views/res_config_settings_views.xml]: string="Expense Product"

File: views/res_config_settings_views.xml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+                                    <field name="bonus_product_id"/>
+                                </div>
+                                <div class="row">
+                                    <label string="Product Default Bonus Rate" for="bonus_rate" class="col-md-3 o_light_label"/>
```
Comment (sbuhl) [inline views/res_config_settings_views.xml]: string="Default Technician Expense Rate"

File: views/product_views.xml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="product_template_form_view_invoice_policy_inherit_gse_bonuses" model="ir.ui.view">
+        <field name="name">product.template.inherit.gse.bonuses.form</field>
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="sale.product_template_form_view"/>
+        <field name="arch" type="xml">
+            <field name="service_tracking" position="after">
+                <field name="bonus_rate" attrs="{'invisible': [('type', '!=', 'service')]}"/>
```
Comment (sbuhl) [inline views/product_views.xml]: @shortcut1337  ceci ne permet-il pas de spécifier un rate spécifique par produit? 
Est-ce que cela ne résoud pas le "truc chiant" https://discord.com/channels/@me/695234725605802044/1177182352364011570 ?

Reword: string="Default Technician Expense Rate"

File: views/bonus_views.xml [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        <field name="res_model">account.move</field>
+        <field name="view_mode">tree,form,pivot,graph</field>
+        <field name="context">{
+            'create': False,
+        }</field>
+    </record>
+
+    <!-- Menu Items in Sales (sale_management) and Field Services (industry_fsm) -->
+    <menuitem id="menu_bonus_sale"
+        name="Bonuses"
```
Comment (sbuhl) [inline views/bonus_views.xml]: Reword: "Technicians"

File: views/bonus_views.xml [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        parent="sale.sale_menu_root"
+        sequence="45"
+        action="gse_bonuses.action_gse_bonus"
+        />
+    <menuitem id="menu_bonus_fsm"
+        name="Bonuses"
```
Comment (sbuhl) [inline views/bonus_views.xml]: Reword: "Technicians"

File: security/security.xml [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <!-- Bonus Security Rules -->
+    <record id="gse_bonus_rule" model="ir.rule">
+        <field name="name">Bonus User</field>
```
Comment (sbuhl) [inline security/security.xml]: name: Technician Transport Expenses User

File: security/security.xml [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+        <field name="model_id" ref="model_gse_bonus"/>
+        <field name="domain_force">[('employee_id', '=', user.employee_id.id)]</field>
+        <field name="groups" eval="[(4, ref('base.group_user'))]"/>
+    </record>
+    <record id="gse_bonus_rule_manager" model="ir.rule">
+        <field name="name">Bonus Manager</field>
```
Comment (sbuhl) [inline security/security.xml]: name: Technician Transport Expenses Manager

File: models/bonus.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+            logger.info("Impossible de générer un bonus pour une SO validée avant le 1er juin 2023 (SO %s %s).", order.id, order.date_order)
+            return
+
+        order.ensure_one()
+
+        # Check if bonuses can be created: the SO must:
+        # - be fully paid
+        # - be fully delivered
+        # - have its related labor service tasks done (related SOL will be
+        #   marked as delivered, so it should be covered by previous point)
```
Comment (sbuhl) [inline models/bonus.py:93]: Can we add 2 fields and 2 conditions that have just emerged:
1. On an employee's contract, add a boolean: Allow Technician Transport Expenses.
2. On the task, add a boolean: Exclude from Technician Transport Expenses.
If the first is false, do not compute the bonus
If the second is true, do not compute the bonus.

Comment (sbuhl) [reply models/bonus.py:93]: Screenshot 2023-12-21 at 14 01 51
Screenshot 2023-12-21 at 14 04 01

**Review by sbuhl — COMMENTED** [review]
Hello @shortcut1337 
Thank you for the commits. 
I've 2 main requests before merge.
1. Few wording (we don't want to see "bonuses" in the interface but "transport expense"
2. Add 2 new conditions before creating a bonus. 

Do you think it's feasible in a relatively short timeframe, or should we go ahead and merge as is and make the changes afterwards?

File: views/product_views.xml (hunk 9 above)
Comment (shortcut1337) [reply views/product_views.xml]: C'est pas si simple, car `generate_bonuses()` ne donne un bonus que:
1. Si le produit est un produit de type service et qui génère une tâche (`service_tracking` field)
2. Et le bonus est calculé en fonction des timesheets liées à la tâche liée à la sale order line liée au produit

Ce qui était la spec initiale.
C'est pas très dur d'autoriser certains autres produits à générer un bonus, faut juste coder une condition spécial dans generate_bonus pour include un autre moyen d'ouvrir le droit à un bonus, et de coder la facon de générer le bonus pour ces types de produit.

Donc c'est un peu plus que juste mettre une valeur dans ce field

Comment (shortcut1337) [reply views/product_views.xml]: En plus de tout ça, tout le modeling (fields) du `gse.bonus`tourne autour d'une timesheet.
Un bonus est otujours lié à une timesheet, c'est grace à la timesheet qu'on retrouve (computed fields):
- la SO line et donc la SO
- l'employee id
- la company et donc la currency

C'est en fait la partie la plus ennuyante mais rien d'impossible, je suis content d'avoir écrit des tests car ça peut du coup casser les bonus de timesheet si je change le modeling

Comment (shortcut1337) [reply views/product_views.xml]: Voila c'est implémenté, je te laisserai testé, cf mon msg sur discord

File: models/bonus.py (hunk 14 above)
Comment (shortcut1337) [reply models/bonus.py:93]: Done

**Comment by shortcut1337** [comment]
Done at https://github.com/GoShop-Energy/technicians/pull/1

**Pull request closed without merging**